platform/osgi/platform-osgi-configadmin/src/main/java/org/apache/felix/cm/impl/helper/BaseTracker.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.felix.cm.impl.helper;
+
+import java.security.AccessControlContext;
+import java.security.AccessController;
+import java.security.DomainCombiner;
+import java.security.Permission;
+import java.security.ProtectionDomain;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Dictionary;
+import java.util.List;
+import org.apache.felix.cm.impl.CaseInsensitiveDictionary;
+import org.apache.felix.cm.impl.ConfigurationManager;
+import org.apache.felix.cm.impl.RankingComparator;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.Constants;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.cm.ConfigurationException;
+import org.osgi.service.cm.ManagedService;
+import org.osgi.service.cm.ManagedServiceFactory;
+import org.osgi.service.log.LogService;
+import org.osgi.util.tracker.ServiceTracker;
+
+/**
+ * The <code>BaseTracker</code> is the base class for tracking <code>ManagedService</code> and
+ * <code>ManagedServiceFactory</code> services. It maps their <code>ServiceRegistration</code> to
+ * the {@link ConfigurationMap} mapping their service PIDs to provided configuration.
+ */
+public abstract class BaseTracker<S> extends ServiceTracker<S, ConfigurationMap<?>> {
```

Comment:
✏️ It would help in reviewing and for future upgrades if the changes to this file from its base state were commented.

Reply:
❓ What did change?